File: LeetCode/Kai/95UniqueBinarySearchTreesII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<TreeNode> generateTrees(int n) {
        return dfs(1, n);
    }
    //This structure is similar to LeetCode 241: Different Ways to Add Parentheses
    public List<TreeNode> dfs(int left, int right) {
        List<TreeNode> result = new LinkedList<TreeNode>();
        
        // if(left==right) {         //redundant code. This condition has been taken care of in the for loop below
        //     TreeNode node = new TreeNode(left);
        //     result.add(node);
        //     return result;
        // }
        
        if(left>right) {
            TreeNode node = null;
            result.add(node);
            return result;
        }
        
        for(int i=left; i<=right; i++) {
            List<TreeNode> leftTrees = dfs(left, i-1);
            List<TreeNode> rightTrees = dfs(i+1, right);
            for(TreeNode leftTree : leftTrees) {
                for(TreeNode rightTree : rightTrees) {
                    TreeNode root = new TreeNode(i);
                    root.left = leftTree;
                    root.right = rightTree;
                    result.add(root);
                }
            }
        }
        
        return result;
    }
    
}

File: LeetCode/Liang/95UniqueBinarySearchTreesII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        vector<TreeNode*> result = generateHelper(1, n);
        return result;
    }
    
private:
    vector<TreeNode*> combineHelper(vector<TreeNode*> &first, vector<TreeNode*> &second, int rootVal) { // Why I cannot use TreeNode *root here as the API?
        vector<TreeNode*> result;
        for (int i = 0; i < first.size(); ++i) {
            for (int j = 0; j < second.size(); ++j) {
                TreeNode *root = new TreeNode(rootVal);
                root->left = first.at(i);
                root->right = second.at(j);
                result.push_back(root);
            }
        }
        return result;
    }
    
    vector<TreeNode*> generateHelper(int begin, int end) {
        if (begin > end) {
            return vector<TreeNode*>(1, nullptr);
        }
        
        if (begin == end) {
            return vector<TreeNode*>(1, new TreeNode(begin));
        }
        
        vector<TreeNode*> result;
        for (int i = begin; i <= end; ++i) {
            vector<TreeNode*> left = generateHelper(begin, i - 1);
            vector<TreeNode*> right = generateHelper(i + 1, end);
            vector<TreeNode*> tmpRes = combineHelper(left, right, i);
            result.insert(result.end(), tmpRes.begin(), tmpRes.end()); // Not result.back();
        }
        return result;
    }
};
